// alsatian/dist/core/alsatian-core.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const _1 = require("./");
exports.TestOutputStream = _1.TestOutputStream;
exports.TestSet = _1.TestSet;
const expect_1 = require("./expect");
exports.buildExpect = expect_1.buildExpect;
exports.Expect = expect_1.Expect;
const matchers_1 = require("./matchers");
exports.ContainerMatcher = matchers_1.ContainerMatcher;
exports.EmptyMatcher = matchers_1.EmptyMatcher;
exports.FunctionMatcher = matchers_1.FunctionMatcher;
exports.FunctionSpyMatcher = matchers_1.FunctionSpyMatcher;
exports.Matcher = matchers_1.Matcher;
exports.NumberMatcher = matchers_1.NumberMatcher;
exports.PropertyMatcher = matchers_1.PropertyMatcher;
exports.StringMatcher = matchers_1.StringMatcher;
const decorators_1 = require("./decorators");
exports.AsyncSetup = decorators_1.AsyncSetup;
exports.AsyncSetupFixture = decorators_1.AsyncSetupFixture;
exports.AsyncTeardown = decorators_1.AsyncTeardown;
exports.AsyncTeardownFixture = decorators_1.AsyncTeardownFixture;
exports.AsyncTest = decorators_1.AsyncTest;
exports.Focus = decorators_1.Focus;
exports.FocusTest = decorators_1.FocusTest;
exports.FocusTests = decorators_1.FocusTests;
exports.Ignore = decorators_1.Ignore;
exports.IgnoreTest = decorators_1.IgnoreTest;
exports.IgnoreTests = decorators_1.IgnoreTests;
exports.Setup = decorators_1.Setup;
exports.SetupFixture = decorators_1.SetupFixture;
exports.Teardown = decorators_1.Teardown;
exports.TeardownFixture = decorators_1.TeardownFixture;
exports.Test = decorators_1.Test;
exports.TestCase = decorators_1.TestCase;
exports.TestCases = decorators_1.TestCases;
exports.TestFixture = decorators_1.TestFixture;
exports.Timeout = decorators_1.Timeout;
const spying_1 = require("./spying");
exports.Any = spying_1.Any;
exports.createFunctionSpy = spying_1.createFunctionSpy;
exports.FunctionSpy = spying_1.FunctionSpy;
exports.RestorableFunctionSpy = spying_1.RestorableFunctionSpy;
exports.SpyOn = spying_1.SpyOn;
exports.SpyOnProperty = spying_1.SpyOnProperty;
const results_1 = require("./results");
exports.TestCaseResult = results_1.TestCaseResult;
exports.TestFixtureResults = results_1.TestFixtureResults;
exports.TestOutcome = results_1.TestOutcome;
exports.TestResults = results_1.TestResults;
exports.TestSetResults = results_1.TestSetResults;
const errors_1 = require("./errors");
exports.MatchError = errors_1.MatchError;
exports.TestTimeoutError = errors_1.TestTimeoutError;
const METADATA_KEYS = require("./decorators/_metadata-keys");
exports.METADATA_KEYS = METADATA_KEYS;
const running_1 = require("./running");
exports.TestRunner = running_1.TestRunner;
const log_1 = require("./maintenance/log");
const log = log_1.Logger.log;
exports.log = log;
